tooling/nargo_cli/src/cli/trace_cmd.rs

```diff
@@ -36,9 +45,76 @@ pub(crate) struct TraceCommand {
     /// Directory where to store trace.json
     #[clap(long, short)]
     trace_dir: String,
+
+    /// Insert plonky2 information to the trace file
+    #[arg(long)]
+    trace_plonky2: bool,
+}
+
+fn generate_plonky2_debug_trace_list(
+    args: TraceCommand,
+    config: NargoConfig,
+) -> Result<DebugTraceList, CliError> {
+    let toml_path = get_package_manifest(&config.program_dir)?;
+    //let default_selection =
+    //    if args.workspace { PackageSelection::All } else { PackageSelection::DefaultOrAll };
+    let selection = args.package.map_or(PackageSelection::DefaultOrAll, PackageSelection::Selected);
+    let workspace = resolve_workspace_from_toml(
+        &toml_path,
+        selection,
+        Some(NOIR_ARTIFACT_VERSION_STRING.to_string()),
+    )?;
+
+    let mut workspace_file_manager = file_manager_with_stdlib(std::path::Path::new(""));
+    insert_all_files_for_workspace_into_file_manager(&workspace, &mut workspace_file_manager);
+    let parsed_files = parse_all(&workspace_file_manager);
+
+    let binary_packages = workspace.into_iter().filter(|package| package.is_binary());
+    for package in binary_packages {
+        let compilation_result = compile_program(
+            &workspace_file_manager,
+            &parsed_files,
+            &workspace,
+            package,
+            &args.compile_options,
+            None,
+            true,
+            true,
+        );
+
+        let dtl = if let Ok((compiled_program, _)) = &compilation_result {
+            if let Some(plonky2_circuit) = &compiled_program.plonky2_circuit {
+                plonky2_circuit.debug_trace_list.clone()
+            } else {
+                None
+            }
+        } else {
+            None
+        };
```

Comment:
Even though at first glance that sounds like a good idea, after some careful thinking I realized why I intentionally wrote it the way I originally did in the first place. Since, it's not obvious, I added comments, explaining it: https://github.com/blocksense-network/noir/pull/96/commits/1aff951d84d9c5c89c6e4c2b54e9137bd73f2ebf

Reply:
The TLDR story is: we want to delay returning an error, until the call to report_errors, because it will better explain to the user *why* the compiler didn't produce a plonky2 output. In fact, logically, we want to initialize the debug_trace_list variable *after* report_errors(), however Rust's borrow checker forces us to do this before report_errors(). So, we set debug_trace_list to None to avoid crashing before report_errors().

Reply:
There are two alternative ways to avoid the borrow checker restrictions and moving the debug_trace_list initialization after report_errors().
1) calling report_errors() with compilation_result.clone(). This doesn't work, because clone() is not implemented for the plonky2 circuit.
2) calling report_errors() with a borrow: &compilation_result. This requires changing the type of the first parameter to report_errors() from CompilationResult<T> to &CompilationResult<T>, which causes a lot of cascading compiler errors and other types that need to be changed as well. Some are not very practical to fix, for example the result of report_errors() must now also contain a & borrow, which now requires a lifetime parameter, etc.
That's why I think my original solution is the least bad one. IMHO, it is also a nice example of how some good developer's time is wasted fighting the borrow checker.

Reply:
> That's why I think my original solution is the least bad one. IMHO, it is also a nice example of how some good developer's time is wasted fighting the borrow checker.

Agreed, although I would beat my favorite dead horse: the upstream noir code 😁 

I think the error here is that report_errors should not consume compilation_result, as it feels like an optional operation. That would mean alternative 2. from your two options should have been the case from the get go.

I agree that the way you've done it is the best, given our poor inheritance.